https://github.com/neovim/neovim/actions/runs/3246436285/jobs/5325170873#step:5:765

```
2022-10-13T23:51:00.0037291Z [315/407] Building C object src\nvim\CMakeFiles\nvim.dir\os\os_win_console.c.obj
2022-10-13T23:51:00.0037737Z D:\a\neovim\neovim\src\nvim\os\os_win_console.c(82): warning C4013: 'os_getenv' undefined; assuming extern returning int
2022-10-13T23:51:00.0038238Z D:\a\neovim\neovim\src\nvim\os\os_win_console.c(85): warning C4013: 'os_path_exists' undefined; assuming extern returning int
2022-10-13T23:51:00.0038762Z D:\a\neovim\neovim\src\nvim\os\os_win_console.c(82): warning C4047: 'initializing': 'const char *' differs in levels of indirection from 'int'
```

--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -211,17 +211,7 @@ HWND hWnd = NULL;
 static HICON hOrigIconSmall = NULL;
 static HICON hOrigIcon = NULL;
 
-/// Save Windows console icon to be reset later
-static void SaveWin32ConsoleIcon(void)
-{
-  if ((hWnd = GetConsoleWindow()) == NULL) {
-    return;
-  }
-  hOrigIconSmall = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_SMALL, (LPARAM)0);
-  hOrigIcon = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_BIG, (LPARAM)0);
-}
-
-static void SetConsoleIcon(HWND hWindow, HICON hIconSmall, HICON hIcon)
+static void os_icon_set(HWND hWindow, HICON hIconSmall, HICON hIcon)
 {
   if (hWindow == NULL) {
     return;
@@ -234,15 +224,18 @@ static void SetConsoleIcon(HWND hWindow, HICON hIconSmall, HICON hIcon)
   }
 }
 
-/// Reset Windows console icon to original
-static void ResetWin32ConsoleIcon(void)
+/// Sets Nvim logo as Windows console icon.
+///
+/// Saves the original icon so it can be restored at exit.
+static void os_icon_init(void)
 {
-  SetConsoleIcon(hWnd, hOrigIconSmall, hOrigIcon);
-}
+  if ((hWnd = GetConsoleWindow()) == NULL) {
+    return;
+  }
+  // Save Windows console icon to be restored later.
+  hOrigIconSmall = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_SMALL, (LPARAM)0);
+  hOrigIcon = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_BIG, (LPARAM)0);
 
-/// Set Neovim logo as Windows console icon
-static void SetWin32ConsoleIcon(void)
-{
   const char *vimruntime = os_getenv("VIMRUNTIME");
   if (vimruntime != NULL) {
     snprintf(NameBuff, MAXPATHL, "%s" _PATHSEPSTR "neovim.ico", vimruntime);
@@ -251,7 +244,7 @@ static void SetWin32ConsoleIcon(void)
     } else {
       HICON hVimIcon = LoadImage(NULL, NameBuff, IMAGE_ICON, 64, 64,
                                  LR_LOADFROMFILE | LR_LOADMAP3DCOLORS);
-      SetConsoleIcon(hWnd, hVimIcon, hVimIcon);
+      os_icon_set(hWnd, hVimIcon, hVimIcon);
     }
   }
 }
@@ -305,11 +298,6 @@ int main(int argc, char **argv)
   event_init();
 
   early_init(&params);
-
-#ifdef MSWIN
-  SaveWin32ConsoleIcon();
-  SetWin32ConsoleIcon();
-#endif
 
   set_argv_var(argv, argc);  // set v:argv
 
@@ -599,6 +587,12 @@ int main(int argc, char **argv)
     TIME_MSG("UIEnter autocommands");
   }
 
+#ifdef MSWIN
+  if (use_builtin_ui) {
+    os_icon_init();
+  }
+#endif
+
   // Adjust default register name for "unnamed" in 'clipboard'. Can only be
   // done after the clipboard is available and all initial commands that may
   // modify the 'clipboard' setting have run; i.e. just before entering the
@@ -776,7 +770,8 @@ void getout(int exitval)
   }
 
 #ifdef MSWIN
-  ResetWin32ConsoleIcon();
+  // Restore Windows console icon before exiting.
+  os_icon_set(hWnd, hOrigIconSmall, hOrigIcon);
 #endif
 
   os_exit(exitval);

--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -206,50 +206,6 @@ void early_init(mparm_T *paramp)
   init_signs();
 }
 
-#ifdef MSWIN
-HWND hWnd = NULL;
-static HICON hOrigIconSmall = NULL;
-static HICON hOrigIcon = NULL;
-
-static void os_icon_set(HWND hWindow, HICON hIconSmall, HICON hIcon)
-{
-  if (hWindow == NULL) {
-    return;
-  }
-  if (hIconSmall != NULL) {
-    SendMessage(hWnd, WM_SETICON, (WPARAM)ICON_SMALL, (LPARAM)hIconSmall);
-  }
-  if (hIcon != NULL) {
-    SendMessage(hWnd, WM_SETICON, (WPARAM)ICON_BIG, (LPARAM)hIcon);
-  }
-}
-
-/// Sets Nvim logo as Windows console icon.
-///
-/// Saves the original icon so it can be restored at exit.
-static void os_icon_init(void)
-{
-  if ((hWnd = GetConsoleWindow()) == NULL) {
-    return;
-  }
-  // Save Windows console icon to be restored later.
-  hOrigIconSmall = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_SMALL, (LPARAM)0);
-  hOrigIcon = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_BIG, (LPARAM)0);
-
-  const char *vimruntime = os_getenv("VIMRUNTIME");
-  if (vimruntime != NULL) {
-    snprintf(NameBuff, MAXPATHL, "%s" _PATHSEPSTR "neovim.ico", vimruntime);
-    if (!os_path_exists(NameBuff)) {
-      WLOG("neovim.ico not found: %s", NameBuff);
-    } else {
-      HICON hVimIcon = LoadImage(NULL, NameBuff, IMAGE_ICON, 64, 64,
-                                 LR_LOADFROMFILE | LR_LOADMAP3DCOLORS);
-      os_icon_set(hWnd, hVimIcon, hVimIcon);
-    }
-  }
-}
-#endif
-
 #ifdef MAKE_LIB
 int nvim_main(int argc, char **argv);  // silence -Wmissing-prototypes
 int nvim_main(int argc, char **argv)
@@ -771,7 +727,7 @@ void getout(int exitval)
 
 #ifdef MSWIN
   // Restore Windows console icon before exiting.
-  os_icon_set(hWnd, hOrigIconSmall, hOrigIcon);
+  os_icon_set(NULL, NULL);
 #endif
 
   os_exit(exitval);

--- a/src/nvim/os/os_win_console.c
+++ b/src/nvim/os/os_win_console.c
@@ -2,12 +2,17 @@
 // it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
 
 #include "nvim/os/input.h"
+#include "nvim/os/os.h"
 #include "nvim/os/os_win_console.h"
 #include "nvim/vim.h"
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "os/os_win_console.c.generated.h"
 #endif
+
+static HWND hWnd = NULL;
+static HICON hOrigIconSmall = NULL;
+static HICON hOrigIcon = NULL;
 
 int os_get_conin_fd(void)
 {
@@ -44,4 +49,46 @@ void os_replace_stdout_and_stderr_to_conout(void)
   close(STDERR_FILENO);
   const int conerr_fd = _open_osfhandle((intptr_t)conout_handle, 0);
   assert(conerr_fd == STDERR_FILENO);
+}
+
+/// Sets Windows console icon, or pass NULL to restore original icon.
+void os_icon_set(HICON hIconSmall, HICON hIcon)
+{
+  if (hWnd == NULL) {
+    return;
+  }
+  hIconSmall = hIconSmall ? hIconSmall : hOrigIconSmall;
+  hIcon = hIcon ? hIcon : hOrigIcon;
+
+  if (hIconSmall != NULL) {
+    SendMessage(hWnd, WM_SETICON, (WPARAM)ICON_SMALL, (LPARAM)hIconSmall);
+  }
+  if (hIcon != NULL) {
+    SendMessage(hWnd, WM_SETICON, (WPARAM)ICON_BIG, (LPARAM)hIcon);
+  }
+}
+
+/// Sets Nvim logo as Windows console icon.
+///
+/// Saves the original icon so it can be restored at exit.
+void os_icon_init(void)
+{
+  if ((hWnd = GetConsoleWindow()) == NULL) {
+    return;
+  }
+  // Save Windows console icon to be restored later.
+  hOrigIconSmall = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_SMALL, (LPARAM)0);
+  hOrigIcon = (HICON)SendMessage(hWnd, WM_GETICON, (WPARAM)ICON_BIG, (LPARAM)0);
+
+  const char *vimruntime = os_getenv("VIMRUNTIME");
+  if (vimruntime != NULL) {
+    snprintf(NameBuff, MAXPATHL, "%s" _PATHSEPSTR "neovim.ico", vimruntime);
+    if (!os_path_exists(NameBuff)) {
+      WLOG("neovim.ico not found: %s", NameBuff);
+    } else {
+      HICON hVimIcon = LoadImage(NULL, NameBuff, IMAGE_ICON, 64, 64,
+                                 LR_LOADFROMFILE | LR_LOADMAP3DCOLORS);
+      os_icon_set(hVimIcon, hVimIcon);
+    }
+  }
 }

